aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1465,4 +1466,76 @@ Tensor& diag_out(Tensor &result, const Tensor& self, int64_t dimension) {
   return result;
 }
 
+template <typename scalar_t>
+void apply_renorm(Tensor &res, const Tensor& src, scalar_t value, int64_t dimension, scalar_t maxnorm) {
+  dimension = at::maybe_wrap_dim(dimension, src);
+  TORCH_CHECK(dimension >= 0 && dimension < src.dim(), "invalid dimension ",
+      dimension);
+  TORCH_CHECK(value > 0, "non-positive-norm not supported");
+  TORCH_CHECK(src.dim() > 1, "need at least 2 dimensions, got ", src.dim(), " dimensions");
+
+  Tensor rowR = at::empty({0}, src.options());
+  Tensor rowS = at::empty({0}, src.options());
+
+  res.resize_as_(src);
+
+  for (int64_t i = 0; i < src.size(dimension); i++) {
+    scalar_t norm = 0;
+    scalar_t new_norm;
+
+    rowS = src.select(dimension, i);
+    rowR = res.select(dimension, i);
+    scalar_t* rowS_data = rowS.data_ptr<scalar_t>();
+    scalar_t* rowR_data = rowR.data_ptr<scalar_t>();
+
+    at::parallel_for(0, rowS.numel(), 1, [&](int64_t j_begin, int64_t j_end) {
+      for(int64_t j = j_begin; j < j_end; j++) {
+        if (value == 1) {
+          norm += std::abs(rowS_data[j]);
+        } else if (value == 2) {
+          scalar_t z = rowS_data[j];
+          norm += z * z;
+        } else if (value == INFINITY) {
+          norm = std::max(norm, std::abs(rowS_data[j]));
+        } else {
+          norm += std::pow(std::abs(rowS_data[j]), value);
+        }
+      }
+    });
+
+    if (value != INFINITY) {
+      norm = std::pow(norm, 1 / value);
+    }
+
+    if (norm > maxnorm) {
+      new_norm = maxnorm / (norm + 1e-7);
+      at::parallel_for(0, rowR.numel(), 1, [&](int64_t j_begin, int64_t j_end) {
+        for(int64_t j = j_begin; j < j_end; j++) {
+          rowR_data[j] = rowS_data[j] * new_norm;
+        }
+      });
+    } else {
+      rowR.copy_(rowS);
+    }
+  }
+}
+
+Tensor renorm(const Tensor& self, Scalar p, int64_t dim, Scalar maxnorm) {
+  Tensor result = at::empty({0}, self.options());
+  at::renorm_out(result, self, p, dim, maxnorm);
+  return result;
+}
+
+Tensor& renorm_(Tensor& self, Scalar p, int64_t dim, Scalar maxnorm) {
+  at::renorm_out(self, self, p, dim, maxnorm);
+  return self;
+}
+
+Tensor& renorm_out(Tensor &result, const Tensor& self, Scalar p, int64_t dim, Scalar maxnorm) {
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "renorm", [&] {
+    apply_renorm<scalar_t>(result.contiguous(), self.contiguous(), p.to<scalar_t>(), dim, maxnorm.to<scalar_t>());
```

Comment:
If you are making self and result contiguous, you will have problems with `x.renorm_`called on non-contiguous tensor. Please add such test and fix it.